prov/sockets/src/sock_progress.c

```diff
@@ -1826,25 +1826,22 @@ static int sock_pe_progress_tx_atomic(struct sock_pe *pe,
 		return 0;
 	len += entry_len;
 
-	/* cmp data */
 	datatype_sz = fi_datatype_size(pe_entry->pe.tx.tx_op.atomic.datatype);
-	for (i = 0; i < pe_entry->pe.tx.tx_op.atomic.cmp_iov_len; i++) {
-		if (sock_pe_send_field(pe_entry,
-				       (void *) (uintptr_t) pe_entry->pe.tx.tx_iov[i].cmp.ioc.addr,
-				       pe_entry->pe.tx.tx_iov[i].cmp.ioc.count *
-				       datatype_sz, len))
-			return 0;
-		len += (pe_entry->pe.tx.tx_iov[i].cmp.ioc.count * datatype_sz);
-	}
-
-	/* data */
 	if (pe_entry->flags & FI_INJECT) {
+		/* data */
 		if (sock_pe_send_field(pe_entry,
 				       &pe_entry->pe.tx.inject[0],
 				       pe_entry->pe.tx.tx_op.src_iov_len, len))
 			return 0;
 		len += pe_entry->pe.tx.tx_op.src_iov_len;
+		/* cmp data */
+		if (sock_pe_send_field(pe_entry,
+					&pe_entry->pe.tx.inject[0] + pe_entry->pe.tx.tx_op.src_iov_len,
+					pe_entry->pe.tx.tx_op.atomic.cmp_iov_len, len))
+			return 0;
+		len += pe_entry->pe.tx.tx_op.atomic.cmp_iov_len;
```

Comment:
This violates the wire protocol. Receiver expects the compare data before the src data. Please check sock_pe_process_rx_atomic().